3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,350 @@
+ * identifiable information) and apply some anonymization (currently only psudonymizantion) on them. Currently a single
+ * strategy for all the fields is supported due to the config format, and there is only one implemented strategy,
+ * however the enrichment supports a strategy per field.
+ *
+ * The user may specify two types of fields POJO or JSON. A POJO field is effectively a scalar field in the
+ * EnrichedEvent, whereas a JSON is a "context" formatted field and it can be wither a scalar in the case of
+ * unstruct_event or an array in the case of derived_events and contexts
+ *
+ * @param fieldList a list of configured PiiFields
+ */
+case class PiiPseudonymizerEnrichment(fieldList: List[PiiField]) extends Enrichment {
+  def transformer(event: EnrichedEvent): Unit = fieldList.foreach(_.transform(event))
+}
+
+/**
+ * Specifies a scalar field in POJO and the strategy that should be applied to it.
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator where the strategy will be applied
+ */
+final case class PiiScalar(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = if (fieldValue != null) strategy.scramble(fieldValue) else null
+}
+
+/**
+ * Specifies a strategy to use, a field mutator where the JSON can be found in the EnrichedEvent POJO, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a json path within the contexts where this strategy will
+ * be applied (the path may correspond to multiple fields).
+ *
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator for the json field
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator, schemaCriterion: SchemaCriterion, jsonPath: String)
+    extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String): String =
+    if (fieldValue != null)
+      compact(render(parse(fieldValue).transformField {
+        case JField("data", JArray(contexts)) =>
+          ("data", JArray(contexts.map {
+            case JObject(context) => modifyObjectIfSchemaMatches(context)
+            case x                => x
+          }))
+        case JField("data", JObject(unstructEvent)) =>
+          ("data", modifyObjectIfSchemaMatches(unstructEvent))
+      }))
+    else null
+
+  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)]): JObject = {
+    val fields: Map[String, JValue] = context.toMap.filterKeys(s => s == "schema" || s == "data")
+    fields
+      .get("schema")
+      .flatMap(s =>
+        SchemaKey.parse(s.extract[String]).map(schemaCriterion.matches).toOption match {
```

Comment:
Having recursive nature of `transformField` it turns out that if we'll get a context like:

```
{
  "schema": "iglu:com.acme/event/jsonschema/1-0-0",
  "data": {"schema": 1, "data": 2, "foo": 3}
}
```

then `s.extract[String]` will fail at runtime? If that is the case (I'd like to have a test for that assumption) then above request about of changing `transformWith` is even stronger.